README
NAME
    Perinci::CmdLine - Rinci/Riap-based command-line application framework

VERSION
    This document describes version 1.43 of Perinci::CmdLine (from Perl
    distribution Perinci-CmdLine), released on 2015-01-07.

SYNOPSIS
    See Perinci::CmdLine::Manual::Examples.

DESCRIPTION
    See Perinci::CmdLine::Manual.

    Screenshots:

REQUEST KEYS
    See also Perinci::CmdLine::Base. Extra stuffs put by this module to the
    $r hash/stash.

    *   format_options => hash

ATTRIBUTES
    All the attributes of Perinci::CmdLine::Base, plus:

  log => BOOL (default: 1)
    Whether to load Log::Any::App (enable logging output) by default. See
    "LOGGING" for more details.

  use_utf8 => BOOL
    From Term::App::Role::Attrs (please see its docs for more details).
    There are several other attributes added by the role.

  undo => BOOL (optional, default 0)
    Whether to enable undo/redo functionality. Some things to note if you
    intend to use undo:

    *   These common command-line options will be recognized

        "--undo", "--redo", "--history", "--clear-history".

    *   Transactions will be used

        "use_tx=>1" will be passed to Perinci::Access, which will cause it
        to initialize the transaction manager. Riap requests begin_tx and
        commit_tx will enclose the call request to function.

    *   Called function will need to support transaction and undo

        Function which does not meet qualifications will refuse to be
        called.

        Exception is when subcommand is specified with "undo=>0", where
        transaction will not be used for that subcommand. For an example of
        disabling transaction for some subcommands, see "bin/u-trash" in the
        distribution.

  undo_dir => STR (optional, default ~/.<program_name>/.undo)
    Where to put undo data. This is actually the transaction manager's data
    dir.

METHODS
    All the methods of Perinci::CmdLine::Base, plus:

RESULT METADATA
    All those supported by Perinci::CmdLine::Base, plus:

  x.hint.result_binary => bool
    If set to true, then when formatting to "text" formats, this class won't
    print any newline to keep the data being printed unmodified.

ENVIRONMENT
    All the environment variables that Perinci::CmdLine::Base supports,
    plus:

  PERINCI_CMDLINE_COLOR_THEME => STR
    Can be used to set "color_theme".

  PROGRESS => BOOL
    Explicitly turn the progress bar on/off.

  COLOR => INT
    Please see Term::App::Role::Attrs.

  UTF8 => BOOL
    Please see Term::App::Role::Attrs.

SEE ALSO
    Perinci, Rinci, Riap.

    Perinci::CmdLine::Base.

    Perinci::CmdLine::Lite.

    Other CPAN modules to write command-line applications: App::Cmd,
    App::Rad, MooseX::Getopt.

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Perinci-CmdLine>.

SOURCE
    Source repository is at
    <https://github.com/perlancar/perl-Perinci-CmdLine>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Perinci-CmdLine>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    perlancar <perlancar@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by perlancar@cpan.org.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perinci-CmdLine

1.42    2015-01-04 (PERLANCAR)

	- No functional changes.

	- Remove unneeded dep [RT#101270].


1.41    2014-12-19 (PERLANCAR)

        - Adjust to new pericmd-base (0.68, refactored common bits of default
          common_opts to ::Base to reduce code duplication).


1.40    2014-12-11 (PERLANCAR)

        - Use new name of renamed modules: SHARYANTO::Role::ColorTheme ->
          Color::Theme::Role::ANSI, SHARYANTO::Role::TermAttrs ->
          Term::App::Role::Attrs.


1.39    2014-12-02 (PERLANCAR)

	- No functional changes.

	- Switch Perinci::Sub::To::{CLIOptSpec,CLIDocData}.


1.38    2014-11-21 (PERLANCAR)

        - Use Perinci::Sub::To::CLIOptSpec to avoid code duplication and produce
          nicer list of options for displaying in --help.

        - Tweak --help output.

        [REMOVED FEATURES]

        - Like in Perinci::CmdLine::Lite, now only provide --no-config instead
          of also --noconfig.


1.37    2014-11-09 (PERLANCAR)

	- [Incompatible change] Rename attribute 'log_any_app' to 'log' for
	  compatibility with Perinci::CmdLine::Lite, old name is still
	  provided for backward compatibility reason but will be removed in
	  the future.

	- Document about x.hint.result_binary.


1.36    2014-10-24 (PERLANCAR)

	- Update dep version (has better binary result/arg support).

	- Format binary result as-is (no newline/send to
	  Perinci::Result::Format) when format is text.


1.35    2014-10-24 (PERLANCAR)

	- Default utf8 to off if outputting binary data.

	- Add attribute: riap_version (and bump Riap client deps with
	  versions which already support riap_version).

	- Typo: unimplemented status is 501, not 502.


1.34    2014-10-14 (PERLANCAR)

	- No functional changes.

	- Include more screenshots.


1.33    2014-10-14 (PERLANCAR)

	- No functional changes.

	- Include screenshots.


1.32    2014-10-03 (PERLANCAR)

	- Avoid error when completing (no use_utf8 method).


1.31    2014-09-16 (PERLANCAR)

	- No functional changes.

        - Move page/stream output functionality to Perinci::CmdLine::Base.


1.30    2014-09-11 (PERLANCAR)

	- Add support for cmdline_src=stdin_line and is_password (Rinci
	  1.1.53).


1.29     2014-09-05 (PERLANCAR)

         - [Bugfix] Forgot another json decode which should've been encode.


1.28     2014-09-04 (PERLANCAR)

         - [Bugfix] Fix bug in displaying 'in' schema property in help.


1.27     2014-09-04 (PERLANCAR)

         - [Bugfix] Fix bugs that caused App::riap to die.


1.26     2014-09-03 (PERLANCAR)

         - No functional changes.

         - Update to Perinci::CmdLine::Base 0.26.


1.25     2014-09-02 (SHARYANTO)

         - [Bugfix] Help on subcommand produced main help (some old code has not
           been adjusted).


1.24     2014-08-27 (SHARYANTO)

         - No functional changes.

         - [Internal] Move cmdline_src handling functionality to
           Perinci::CmdLine::Base.


1.23     2014-08-26 (SHARYANTO)

         - No functional changes.

         - Use Perinci::CmdLine::Base 0.22 (change of configuration profile
           syntax).


1.22     2014-08-26 (SHARYANTO)

         - No functional changes.

         - Update translations (id, fr).

         - Fix tests so non-English locale can pass too [CT].


1.21     2014-08-25 (SHARYANTO)

         [ENHANCEMENTS]

         - Add support for reading configuration file.


1.20     2014-08-24 (SHARYANTO)

         - [Bug fix] Edit to make tx-related actions (--undo, --redo, --history,
           etc) work.


1.19     2014-08-23 (SHARYANTO)

         [INCOMPATIBLE CHANGES]

         - Replace result metadata cmdline.display_result with cmdline.result.

         [ENHANCEMENTS]

         - Add result metadata cmdline.skip_format.


1.18     2014-08-22 (SHARYANTO)

         - [Bugfix] Try to init attributes in BUILD again (previously this is
           moved to hook_before_run because of some weirdness that makes
           accessing attribute from role not work in inner block of BUILD).


1.17     2014-08-22 (SHARYANTO)

         [REMOVED FEATURES]

         - Remove scripts strip-result-env-* (users can use 'jq' or other tools
           to do this).

         - Move script peri-run to Perinci-CmdLine-Any.

         - Remove --action (see note in the Incompatible Changes section about
           the change in the way command-line options are parsed).

         - Move examples/ to Perinci-CmdLine-Examples, to help reduce
           dependencies.

         [INCOMPATIBLE CHANGES]

         - To reduce confusion and make things simpler, change the way
           command-line options are parsed. Argument options which happen to
           clash with common options (e.g. --help, --version) are now renamed to
           --help-arg (or --help-arg2 and so on) so no options clash with each
           other. Thus, --action is no longer needed.

         [INTERNAL]

         - Refactoring: split some functionality to Perinci::CmdLine::Base (in
           separate dist) for creating Perinci::CmdLine::Lite, a lightweight
           alternative that doesn't use Moo, Perinci::Access, nor
           Perinci::Result::Format. There are now Perinci-CmdLine and
           Perinci-CmdLine-Lite distributions.

         - Refactoring: run() and run_*() now returns enveloped response instead
           of int. All output is now returned as string by run_*() instead of
           printed directly (except for streaming output).

         - Refactoring: No long uses internal attributes like
           $self->{_subcommand}. Instead, now a per-request (per-run())
           stash/hash $r is passed around. This is not unlike the technique used
           in Apache handlers and Data::Sah. It's cleaner because the
           per-request state is isolated from the object itself.

         - Refactoring: help-generating methods now separated into a separate
           role (Perinci::CmdLine::Role::Help), just to be more organized.


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.025.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Rinci/Riap-based command-line application framework",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06",
    "Perl::osnames" => "0.09"
  },
  "DISTNAME" => "Perinci-CmdLine",
  "EXE_FILES" => [
    "bin/peri-func-usage"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Perinci::CmdLine",
  "PREREQ_PM" => {
    "Color::Theme::Role::ANSI" => 0,
    "Data::Format::Pretty::Console" => 0,
    "Data::Unixish" => "1.39",
    "File::HomeDir" => 0,
    "JSON" => 0,
    "Locale::TextDomain::UTF8" => 0,
    "Log::Any" => 0,
    "Log::Any::App" => 0,
    "Monkey::Patch::Action" => 0,
    "Moo" => 0,
    "Moo::Role" => 0,
    "Perinci::Access" => "0.38",
    "Perinci::Access::HTTP::Client" => "0.21",
    "Perinci::Access::Perl" => "0.74",
    "Perinci::Access::Schemeless" => "0.74",
    "Perinci::Access::Simple::Client" => "0.18",
    "Perinci::CmdLine::Base" => "0.68",
    "Perinci::Object" => "0.11",
    "Perinci::Result::Format" => "0.33",
    "Perinci::Sub::To::CLIDocData" => 0,
    "Perinci::Sub::Wrapper" => "0.40",
    "Progress::Any" => "0.13",
    "Progress::Any::Output" => 0,
    "Progress::Any::Output::TermProgressBarColor" => "0.08",
    "Rinci" => "1.1.41",
    "Term::App::Role::Attrs" => 0,
    "Text::ANSITable" => "0.23",
    "UUID::Random" => 0,
    "experimental" => 0
  },
  "TEST_REQUIRES" => {
    "File::Slurp::Tiny" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Perinci::Examples" => "0.32",
    "Test::More" => "0.98",
    "Test::Perinci::CmdLine" => 0
  },
  "VERSION" => "1.43",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Color::Theme::Role::ANSI" => 0,
  "Data::Format::Pretty::Console" => 0,
  "Data::Unixish" => "1.39",
  "ExtUtils::MakeMaker" => 0,
  "File::HomeDir" => 0,
  "File::ShareDir::Install" => "0.06",
  "File::Slurp::Tiny" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON" => 0,
  "Locale::TextDomain::UTF8" => 0,
  "Log::Any" => 0,
  "Log::Any::App" => 0,
  "Monkey::Patch::Action" => 0,
  "Moo" => 0,
  "Moo::Role" => 0,
  "Perinci::Access" => "0.38",
  "Perinci::Access::HTTP::Client" => "0.21",
  "Perinci::Access::Perl" => "0.74",
  "Perinci::Access::Schemeless" => "0.74",
  "Perinci::Access::Simple::Client" => "0.18",
  "Perinci::CmdLine::Base" => "0.68",
  "Perinci::Examples" => "0.32",
  "Perinci::Object" => "0.11",
  "Perinci::Result::Format" => "0.33",
  "Perinci::Sub::To::CLIDocData" => 0,
  "Perinci::Sub::Wrapper" => "0.40",
  "Perl::osnames" => "0.09",
  "Progress::Any" => "0.13",
  "Progress::Any::Output" => 0,
  "Progress::Any::Output::TermProgressBarColor" => "0.08",
  "Rinci" => "1.1.41",
  "Term::App::Role::Attrs" => 0,
  "Test::More" => "0.98",
  "Test::Perinci::CmdLine" => 0,
  "Text::ANSITable" => "0.23",
  "UUID::Random" => 0,
  "experimental" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}



# modify generated Makefile to generate shell completion scripts. this piece
# is generated by Dist::Zilla::Plugin::GenShellCompletion version 0.07.
GEN_SHELL_COMPLETION:
{
    use Perl::osnames 0.09 qw(is_posix);
    last unless is_posix();

    print "Modifying Makefile to generate shell completion on install\n";
    open my($fh), "<", "Makefile" or die "Can't open generated Makefile: $!";
    my $content = do { local $/; ~~<$fh> };

    $content =~ s/^(install :: pure_install doc_install)/$1 comp_install/m
        or die "Can't find pattern in Makefile (1)";

    $content =~ s/^(uninstall :: .+)/$1 comp_uninstall/m
        or die "Can't find pattern in Makefile (2)";

    $content .= qq|\ncomp_install :\n\t| .
        q|$(PERLRUN) -E'if(eval { require App::shcompgen; 1 }) { system "shcompgen", "--verbose", "generate", "--replace", @ARGV }' -- $(EXE_FILES)| .
        qq|\n\n|;

    $content .= qq|\ncomp_uninstall :\n\t| .
        q|$(PERLRUN) -E'if(eval { require App::shcompgen; 1 }) { system "shcompgen", "--verbose", "remove", @ARGV }' -- $(EXE_FILES)| .
        qq|\n\n|;

    open $fh, ">", "Makefile" or die "Can't write modified Makefile: $!";
    print $fh $content;
}
